apps/builder/app/canvas/features/webstudio-component/webstudio-component.tsx

```diff
@@ -108,12 +119,81 @@ type WebstudioComponentDevProps = {
   components: Components;
 };
 
-export const WebstudioComponentDev = ({
-  instance,
-  instanceSelector,
-  children,
-  components,
-}: WebstudioComponentDevProps) => {
+/**
+ * Radix's VisuallyHiddenPrimitive.Root https://github.com/radix-ui/primitives/blob/main/packages/react/visually-hidden/src/VisuallyHidden.tsx
+ * component makes content from hidden elements accessible to screen readers.
+ * Using the same Content children, however, leads to duplicated React elements, breaking our 'isSelected' logic.
+ * To prevent this, we check if an instance is a descendant of VisuallyHiddenPrimitive.Root, and if so,
+ * we avoid rendering it.
+ */
+const useIsScreenReaderDescendant = (ref: RefObject<HTMLElement>) => {
+  const [isScreenReaderDescendant, setIsScreenReaderDescendant] =
+    useState(false);
+
+  useEffect(() => {
+    if (ref.current !== null) {
+      for (
+        let element: HTMLElement | null = ref.current;
+        element !== null;
+        element = element.parentElement
+      ) {
+        if (
+          element.style.overflow === "hidden" &&
+          element.style.clip === "rect(0px, 0px, 0px, 0px)" &&
+          element.style.position === "absolute" &&
+          element.style.width === "1px" &&
+          element.style.height === "1px"
+        ) {
+          setIsScreenReaderDescendant(true);
+          return;
+        }
+      }
+    }
+  }, [ref]);
+
+  return isScreenReaderDescendant;
+};
+
+/**
+ * For some components that are wrapped with Radix Slot components (where asChild=true),
+ * events are passed implicitly. We aim to merge these implicit events with the explicitly defined ones.
+ **/
+type ImplicitEvents = {
+  onClick?: undefined | ((event: never) => void);
+  onSubmit?: undefined | ((event: never) => void);
+  /**
+   * We ignore the remaining events because we currently detect events using the 'on' prefix.
+   * This approach (defining type like above instead of Partial<Record<`on${string}`, Handler>>)
+   * is necessary due to our TypeScript settings, where 'no exactOptionalPropertyTypes' is set,
+   * which makes it impossible to define a Partial<Record<>> with optional keys.
+   *  (without exactOptionalPropertyTypes ts defines it as {[key: string]: Handler | undefined)}
+   *   instead of {[key: string]?: Handler | undefined)})
+   **/
+};
+
+const existingElements = new Set<string>();
```

Comment:
is this going to grow forever?

Reply:
No, there is delete in effect unsubscribe.